### PR TITLE
XGBoost auto train

### PR DESCRIPTION
**Pull request by yinlou**
1. Automatically tune depth and learning rate or depth only.
2. Automatically rewrite some parameters in auto train if not specified by user.
3. Fix some unit test bugs.

@sperlingxx

### CODE REVIEW

File: python-package/xgboost/training.py [hunk 1]
```diff
@@ -12,6 +12,159 @@
...
+    if p * col_sample_bytree > 2000:
+        params['colsample_bytree'] = 2000.0 / p
+        rabit.tracker_print("[ParamRewrite] colsample_bytree = {}\n".format(params['colsample_bytree']))
+
+    # When # of points are too large, automatically switch to stochastic
+    # gradient boosting.
+    subsample = params.get('subsample', 1)
+    if n * subsample > 1e8:
+        params['subsample'] = 1e6 * 1.0 / n
+        rabit.tracker_print("[ParamRewrite] subsample = {}\n".format(params['subsample']))
```
Comment (sperlingxx) [inline python-package/xgboost/training.py]: I think we only need do tracker_print on one worker (usually which rank = 0).

**Review by sperlingxx — APPROVED** [review]
LGTM
